No support for dev version of cocotb

Hi,
I'm forced to use a bleeding edge version of cocotb because it contains a fix I need for my simulator (Riviera-PRO).

However, pyuvm is currently crashing when I try to use the latest version of cocotb.

Are you in the process of anticipating the upgrade to cocotb 2.0 (which might come soon) so that pyuvm can be used swiftly when its released? In case, I wonder if it could be available through a branch?

Here's a minimal example for how to reproduce issue:

```
import pyuvm

@pyuvm.test()
class DatapathTest(uvm_test):
    """Test ALU with random and max values"""

    pass
```

My environment:

```
pip install cocotb@git+https://github.com/cocotb/cocotb.git@master
...
Successfully installed cocotb-2.0.0.dev0+ca40cae4


pip install pyuvm --no-deps # no-deps required to avoid downgrade of cocotb
...
Successfully installed pyuvm-3.0.0
```

Running test:

```
# COUT:      0.00ns INFO     cocotb                             Running on Riviera-PRO version 2023.10.106.9105
# COUT:      0.00ns INFO     cocotb                             Running tests with cocotb v2.0.0.dev0+ca40cae4 from /home/ogrottvi/mosaix_testing/firmware/mosaix_test_system/.venv/lib64/python3.11/site-packages/cocotb
...
# COUT: Traceback (most recent call last):
# COUT:   File "/home/ogrottvi/mosaix_testing/firmware/mosaix_test_system/.venv/lib64/python3.11/site-packages/pygpi/entry.py", line 37, in load_entry
# COUT:     entry_func(argv)
# COUT:   File "/home/ogrottvi/mosaix_testing/firmware/mosaix_test_system/.venv/lib64/python3.11/site-packages/cocotb/__init__.py", line 298, in _initialise_testbench
# COUT:     _setup_regression_manager()
# COUT:   File "/home/ogrottvi/mosaix_testing/firmware/mosaix_test_system/.venv/lib64/python3.11/site-packages/cocotb/__init__.py", line 479, in _setup_regression_manager
# COUT:     regression_manager.discover_tests(*modules)
# COUT:   File "/home/ogrottvi/mosaix_testing/firmware/mosaix_test_system/.venv/lib64/python3.11/site-packages/cocotb/regression.py", line 291, in discover_tests
# COUT:     mod = import_module(module_name)
# COUT:           ^^^^^^^^^^^^^^^^^^^^^^^^^^
# COUT:   File "/usr/lib64/python3.11/importlib/__init__.py", line 126, in import_module
# COUT:     return _bootstrap._gcd_import(name[level:], package, level)
# COUT:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
# COUT:   File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
# COUT:   File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
# COUT:   File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
# COUT:   File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
# COUT:   File "<frozen importlib._bootstrap_external>", line 940, in exec_module
# COUT:   File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
# COUT:   File "/home/ogrottvi/mosaix_testing/firmware/mosaix_test_system/benches/datapath/datapath_tb.py", line 9, in <module>
# COUT:     import pyuvm
# COUT:   File "/home/ogrottvi/mosaix_testing/firmware/mosaix_test_system/.venv/lib64/python3.11/site-packages/pyuvm/__init__.py", line 8, in <module>
# COUT:     from pyuvm.s06_reporting_classes import *
# COUT:   File "/home/ogrottvi/mosaix_testing/firmware/mosaix_test_system/.venv/lib64/python3.11/site-packages/pyuvm/s06_reporting_classes.py", line 12, in <module>
# COUT:     from cocotb.log import SimTimeContextFilter
# COUT: ModuleNotFoundError: No module named 'cocotb.log'
# COUT: Traceback (most recent call last):
# COUT:   File "/home/ogrottvi/mosaix_testing/firmware/mosaix_test_system/.venv/lib64/python3.11/site-packages/cocotb/__init__.py", line 313, in _sim_event
# COUT:     regression_manager._fail_simulation(msg)
# COUT:   File "/home/ogrottvi/mosaix_testing/firmware/mosaix_test_system/.venv/lib64/python3.11/site-packages/cocotb/regression.py", line 964, in _fail_simulation
# COUT:     self._abort_test(self._sim_failure)
# COUT:   File "/home/ogrottvi/mosaix_testing/firmware/mosaix_test_system/.venv/lib64/python3.11/site-packages/cocotb/regression.py", line 977, in _abort_test
# COUT:     if self._test_outcome is not None:  # pragma: no cover
# COUT:        ^^^^^^^^^^^^^^^^^^
# COUT: AttributeError: 'RegressionManager' object has no attribute '_test_outcome'
# COUT:      0.00ns ERROR    gpi                                Passing event to upper layer failed
# KERNEL: stopped at time: 0 ps
# VSIM: Simulation has finished. There are no more test vectors to simulate.
endsim
# VSIM: Simulation has finished.
exit
make[1]: *** [/home/ogrottvi/mosaix_testing/firmware/mosaix_test_system/.venv/lib/python3.11/site-packages/cocotb_tools/makefiles/simulators/Makefile.riviera:170: results.xml] Error 1
```

--- a/examples/TinyALU/testbench.py
+++ b/examples/TinyALU/testbench.py
@@ -256,9 +256,8 @@ class AluEnv(uvm_env):
         self.driver.ap.connect(self.scoreboard.result_export)
 
 
-@pyuvm.test()
-class AluTest(uvm_test):
-    """Test ALU with random and max values"""
+class AluTestBase(uvm_test):
+    """Base class for ALU tests with random and max values"""
 
     def build_phase(self):
         self.env = AluEnv("env", self)
@@ -271,9 +270,13 @@ class AluTest(uvm_test):
         await self.test_all.start()
         self.drop_objection()
 
+@pyuvm.test()
+class AluTest(AluTestBase):
+    """Test ALU with random and max values"""
+
 
 @pyuvm.test()
-class ParallelTest(AluTest):
+class ParallelTest(AluTestBase):
     """Test ALU random and max forked"""
 
     def build_phase(self):
@@ -282,7 +285,7 @@ class ParallelTest(AluTest):
 
 
 @pyuvm.test()
-class FibonacciTest(AluTest):
+class FibonacciTest(AluTestBase):
     """Run Fibonacci program"""
 
     def build_phase(self):
@@ -292,7 +295,7 @@ class FibonacciTest(AluTest):
 
 
 @pyuvm.test(expect_fail=True)
-class AluTestErrors(AluTest):
+class AluTestErrors(AluTestBase):
     """Test ALU with errors on all operations"""
 
     def start_of_simulation_phase(self):

--- a/examples/TinyALU_reg/testbench.py
+++ b/examples/TinyALU_reg/testbench.py
@@ -29,6 +29,9 @@ from pathlib import Path
 sys.path.append(str(Path("..").resolve()))
 from tinyalu_utils import TinyAluBfm, Ops, alu_prediction  # noqa: E402
 
+import os
+LANGUAGE = os.getenv("TOPLEVEL_LANG", "verilog")
+
 ##############################################################################
 # TESTS ENTIRE RAL of an ALU
 # The ALU has 2 SRC input operands stored into 2 flops
@@ -575,10 +578,8 @@ class AluEnv(uvm_env):
 # TESTS
 ##############################################################################
 
-
-@pyuvm.test()
-class AluTest(uvm_test):
-    """Test ALU with random and max values"""
+class AluTestBase(uvm_test):
+    """Base class for ALU tests with random and max values"""
 
     def build_phase(self):
         self.env = AluEnv("env", self)
@@ -589,7 +590,7 @@ class AluTest(uvm_test):
     async def run_phase(self):
         self.raise_objection()
 
-        if cocotb.LANGUAGE == "verilog":
+        if LANGUAGE == "verilog":
             # Start clock
             clock = Clock(cocotb.top.clk, 1, units="ns")
             cocotb.start_soon(clock.start())
@@ -597,9 +598,13 @@ class AluTest(uvm_test):
         await self.test_all.start()
         self.drop_objection()
 
+@pyuvm.test()
+class AluTest(AluTestBase):
+    """Test ALU with random and max values"""
+
 
 # @pyuvm.test()
-# class ParallelTest(AluTest):
+# class ParallelTest(AluTestBase):
 #     """Test ALU random and max forked"""
 
 #     def build_phase(self):
@@ -608,7 +613,7 @@ class AluTest(uvm_test):
 
 
 @pyuvm.test()
-class FibonacciTest(AluTest):
+class FibonacciTest(AluTestBase):
     """Run Fibonacci program"""
 
     def build_phase(self):
@@ -618,7 +623,7 @@ class FibonacciTest(AluTest):
 
 
 @pyuvm.test()
-class AluTestErrors(AluTest):
+class AluTestErrors(AluTestBase):
     """Test ALU with errors on all operations"""
 
     def build_phase(self):

--- a/pyuvm/extension_classes.py
+++ b/pyuvm/extension_classes.py
@@ -1,5 +1,4 @@
 import functools
-import inspect
 
 import cocotb
 
@@ -31,18 +30,11 @@ def test(
             stage=stage,
         )
         @functools.wraps(cls)
-        async def test(_):
+        async def test_obj(_):
             await uvm_root().run_test(
                 cls, keep_singletons=keep_singletons, keep_set=keep_set
             )
 
-        if cocotb_version_info < (2, 0):
-            # adds cocotb.test object to caller's module
-            caller_frame = inspect.stack()[1]
-            caller_module = inspect.getmodule(caller_frame[0])
-            setattr(caller_module, f"test_{test._id}", test)
-
-        # returns decorator class unmodified
-        return cls
+        return test_obj
 
     return decorator

--- a/pyuvm/utility_classes.py
+++ b/pyuvm/utility_classes.py
@@ -11,6 +11,13 @@ logging.addLevelName(FIFO_DEBUG, "FIFO_DEBUG")
 logging.addLevelName(PYUVM_DEBUG, "PYUVM_DEBUG")
 
 
+if int(cocotb.__version__.split('.')[0]) >= 2:
+    from cocotb.task import current_task
+else:
+    def current_task():
+        return cocotb.scheduler._current_task
+
+
 def count_bits(nn):
     """
     Count the number of bits in a number
@@ -287,7 +294,7 @@ class UVMQueue(cocotb.queue.Queue):
         """
         while self.empty():
             event = Event('{} peek'.format(type(self).__name__))
-            self._getters.append((event, cocotb.scheduler._current_task))
+            self._getters.append((event, current_task()))
             await event.wait()
         return self.peek_nowait()
 

--- a/tests/cocotb_tests/decorator/test_decorator.py
+++ b/tests/cocotb_tests/decorator/test_decorator.py
@@ -92,8 +92,7 @@ class CheckDatumis442(uvm_test):
         self.raise_objection()
         self.drop_objection()
 
-@pyuvm.test()
-class SetMultipleSingletons(uvm_test):
+class SetMultipleSingletonsBase(uvm_test):
     def build_phase(self):
         self.mysingleton = MySingleton()
         self.mysingleton2 = MySingleton2()
@@ -105,6 +104,10 @@ class SetMultipleSingletons(uvm_test):
         self.mysingleton2.datum = 222
         self.mysingleton3.datum = 333
         self.drop_objection()
+
+@pyuvm.test()
+class SetMultipleSingletons(SetMultipleSingletonsBase):
+    ...
 
 pyuvm.test()
 class CheckMultipleSingletonAreNone(uvm_test):
@@ -123,7 +126,7 @@ class CheckMultipleSingletonAreNone(uvm_test):
         self.drop_objection()
 
 @pyuvm.test()
-class SetMultipleSingletonsAgain(SetMultipleSingletons):
+class SetMultipleSingletonsAgain(SetMultipleSingletonsBase):
     pass
 
 @pyuvm.test(keep_singletons=True)

